### PR TITLE
Add Pub/Sub Lite Support

### PR DESCRIPTION
**Pull request by jacobmims**
This pull request is to add support for the Pub/Sub Lite client library. The work is based on the architecture created by @viniciusccarvalho in his Pub/Sub implementation. The usage is largely the same, with most annotations simply adding `Lite` where applicable (eg., `@PubSubLiteClient` vs `@PubSubClient`). 

Some notable differences are that Lite topics and subscriptions require a location parameter (defaulting to `us-central1-a`) and that projects are referenced by project *number* rather than *id* in the topic and subscription names. The client library handles this conversion, but when providing an FQN they are required.

Configuration support has been provided for all settings available in the client library except for the subscriber's `PartitionAssignmentServiceSettings`, because it is a beta API that is still volatile.

I haven't added to the documentation yet, since I wanted to get feedback on the implementation before I spent time on it. 

Thank you for your time and feedback.

### CODE REVIEW

**Comment by viniciusccarvalho** [comment]
@jacobmims thanks for the contribution, I have a few comments on this, and will defer to @graemerocher to chime in:

Pubsub lite is a different dependency. The jury is out on Google's decision to do so, but I believe we should make it a different module if possible, so that we don't add extra dependency if you chose one version versus the other.

If there are any common code, we could refactor them to a pubsub parent module and move the full and lite version as different modules. I would prefer this model as chances are slim that users will use both.

**Comment by jacobmims** [comment]
> @jacobmims thanks for the contribution, I have a few comments on this, and will defer to @graemerocher to chime in:
> 
> Pubsub lite is a different dependency. The jury is out on Google's decision to do so, but I believe we should make it a different module if possible, so that we don't add extra dependency if you chose one version versus the other.
> 
> If there are any common code, we could refactor them to a pubsub parent module and move the full and lite version as different modules. I would prefer this model as chances are slim that users will use both.

@viniciusccarvalho Thank you for the feedback. That makes sense to me. I kept it isolated in a separate package, so it should be an easy refactor to a different module. The only shared code should be annotations, which would also be a fairly easy refactor to a parent module.

**Comment by viniciusccarvalho** [comment]
@jacobmims So I dove more into the code, and into the GCP lite support (which I have still not used). Addressing the elephant in the room, the Google design is a mess :D, and this is where micronaut can actually help :)

We could make this an abstraction for users, I can imagine the pain that would be if someone needs to migrate from lite -> full or vice-versa and had to change all the imports.

Micronaut can actually aid there, we would need a few things at first:

Topic, Subscription, PubSubClient all become common.
Because we inject `Pubsubclient` as an interface we can have different implementations (full and lite)

But we reuse the Topic/Subscription and any other annotation/interface.

Users could now switch between them by just changing the dependency: micronaut-gcp-pubsub or micronaut-gcp-pubsublite.

So now you don't need a `LiteTopic` or `LiteSubscription`, any conversion between projectId and number can also be done via google cloud libraries and abstracted from the user. 

Let me talk to @graemerocher on this, and perhaps move the common interfaces away first, and we can then merge the changes, which I'll think will pretty much be only the `intercept` part of the module, you won't need to re-create the argument binders for instance.

Thanks for this, it will really move the needle on the GCP support :D

**Comment by jacobmims** [comment]
> Topic, Subscription, PubSubClient all become common.
> Because we inject `Pubsubclient` as an interface we can have different implementations (full and lite)
> 
> But we reuse the Topic/Subscription and any other annotation/interface.
> 
> Users could now switch between them by just changing the dependency: micronaut-gcp-pubsub or micronaut-gcp-pubsublite.
> 
> So now you don't need a `LiteTopic` or `LiteSubscription`, any conversion between projectId and number can also be done via google cloud libraries and abstracted from the user.

That was my initial approach when I started on this project. 

Unfortunately, the underlying client libraries between Lite and regular Pub/Sub are non-trivially different, all the way down the the interfaces that they implement for Publisher and Subscriber and the underlying tech supporting the client libraries. Even though conceptually, they're similar, the implementations are so significantly different that it ended up being a monumental effort to try to conform them to the same interface.  

My goal was to get the pattern to be as close as possible to what was in place for Pub/Sub, but I don't think we can get them to this level of conformity without making a lot of compromises to functionality.

**Comment by viniciusccarvalho** [comment]
I'll look into it, it did not seemed so, because we don't expose them to clients. I don't think it will be that much of work, specially on things like using different annotatios it doesn't look like the right way. The AOP interceptor could  abstract that from clie

**Comment by jacobmims** [comment]
> I'll look into it, it did not seemed so, because we don't expose them to clients. I don't think it will be that much of work, specially on things like using different annotatios it doesn't look like the right way. The AOP interceptor could abstract that from clie

I may have been too in the weeds to see it at the time. I appreciate you looking into it. Hopefully there's an elegant solution and we can do some refactoring to get it down to shared annotations.

**Comment by jacobmims** [comment]
> I'll look into it, it did not seemed so, because we don't expose them to clients. I don't think it will be that much of work, specially on things like using different annotations it doesn't look like the right way. The AOP interceptor could abstract that from client. There's quite a lot of code that is just duplicated, and that will make maintenance harder.
> 
> As an example there are 2 processors `PubSubIntroductionAdvice` that handles producers and `PubSubConsumerAdvice` for MessageListeners.
> 
> We can easily modify those classes to be an AbstractFactory and delegate to the concrete Implementation accordingly. We could have a Default Implementation with dependency on the DefaultFactory and using the "full" client library and a Lite Implementation, with the Lite Publisher Factory. SerDes seems to be the same, so are the argument binders.
> 
> To choose one vs other we could use annotations properties (`@PubSubClient(mode = LITE | FULL)` and Full could be default. That fixes the producer side of things.
> 
> As per consumers, same approach, We can set it as a property on the PubSubListener and create the corresponding beans.
> 
> A lot of the code can be reuse, but there's a good amount of change needed here.
> 
> I can refactor the module to be ready for this lift and we can then resume the lite version if you think its ok

I think something similar to your suggestions is manageable, but I'm a bit concerned that the configuration will be unwieldy if we go overboard.

Lite and "full" have completely incompatible configuration. Lite is also regional, and that means that the annotations will be have a values that aren't usable in one mode vs the other (eg., having a "location" field on the Topic/Subscription annotation, even though that value would do nothing for the "full" version of Pub/Sub) which makes using the library more confusing. I like the idea of streamlining the implementation and getting rid of the duplication. I'm not sold on reusing *all* of the annotations (particularly for the Client/Listener/Subscription/Topic), simply because those annotations will get messy and ambiguous with unused or unusable properties. 

TL;DR:
+ I like refactoring to separate modules (pubsub, pubsublite, and pubsub-shared to reduce duplication)
+ I like removing the binder and exception class duplication for maintainability, and I think that shouldn't be too hard to manage.
+ I think we can potentially expand on the utility libraries in the "shared" module to reduce code duplication
- I don't think collapsing the Client/Listener/Topic/Subscription annotations are worth the ambiguity and complexity it would create

**Comment by viniciusccarvalho** [comment]
I still don't think you are understanding the tradeoffs here. Having multiple annotations `@Topic` and `@TopicLite` vs having an extra property (BTW we already have a property called endpoint for regional endpoints on pubsub) is absolutely not ambiguity or complexity. Annotations can have extra properties that can be used on use case basis:

I would much rather have this:

```
@PubSubClient
public interface OrderClient {
    @Topic(value = "orders", endpoint = "us-central1-pubsub.googleapis.com:443") 
    void send(Order order, @OrderingKey String key); 
}

@PubSubClient(mode=PubsubClient.LITE)
public interface OrderClient {
    @Topic(value = "orders", region = "us-central1") 
    void send(Order order, @OrderingKey String key); 
}
```

than this:

```
@PubSubClientLite
public interface OrderClient {
    @TopicLite(value = "orders", region = "us-central1") 
    void send(Order order, @OrderingKey String key); 
}
```

And yes, configuration is supposed to be different, that's why we have the ConfigurationProperties:

```yaml
gcp:
  pubsub:
    publisher:
       my-config:
         executor:
    lite:
      publisher:
        mylite-config:
          executor:
```

As it is, my recommendation right now is that we don't merge this code into the repo. And we keep same interfaces to be shared. Another common example I see is that customers will have a bridge need (draining a lite topic into a global topic). We had this issue to deal back at Spring Cloud Stream too, when users started asking for multiple binders (kafka <-> rabbitmq) at the classpath.

**Comment by jacobmims** [comment]
@viniciusccarvalho I understand what you're saying. This may be something of a design preference on my part, but I think it's important to distinguish between the two things as different products, rather than "Lite" being a subset of the global offering (it isn't really, even though the name suggests it). You used the [Kafka](https://micronaut-projects.github.io/micronaut-kafka/latest/guide/)and [RabbitMq]((https://micronaut-projects.github.io/micronaut-rabbitmq/latest/guide/) products as examples, and I think that is a good way to look at it. In Micronaut, these have qualifiers on their annotations for their products, similar to my suggestion. Both Pub/Sub and Pub/Sub Lite are Google publisher / subscriber messaging offerings, and they do share some things which can be refactored to a common dependency, but they have separate client libraries and the interfaces that they implement are separate. 

When you said "the Google design is a mess" at the beginning, you weren't wrong. Pub/Sub "Lite" is a misnomer, as it's not just a slimmer offering of the same product. You can't read from a Lite Topic with a standard Subscription, or vice versa. They're completely separate products from Google Cloud's perspective. That naming choice is questionable, but that's what we have to work with and where I have concerns of confusion from user's of the library. Honestly, even keeping the same annotation names and separating them out by module / package would be sufficient to address my concern, as that is similar to the approach Google has taken with their client libraries for the two products.

With all of that said, if you want to go down the road of putting both products under the same annotation umbrella, I'm willing to work with that to get the work done. I think this will be helpful for people who want to use this product and help drive adoption of Micronaut for GCP developers. Either way, let me know @graemerocher and your's preference for implementation and I can get started on refactoring things around next week.

**Comment by viniciusccarvalho** [comment]
Here's my final 2c on the subject, a disclaimer I work for Google, I get to talk to those guys quite often on api design, and this is something I may be able to discuss soon, not that I have any control, a lot of teams with their own agendas and opinions.

1. The lite library does inherit from some common Interfaces, so the point of them not being similar is just an implementation issue, that's why having common annotations make total sense (Like `@Id` on data annotations, we don't have a `@SQLId`, `@NoSQLId`). Take the Publisher for instance. The current code already creates a `PublisherInterface` which is a common class for both lite and full implementations. This only shows that we really should use the same common annotations. Depending on how the user configures its classpath we can provide defaults, or users can explicitly  declare their choice. Same approach micronaut-data uses, or for that matter the `HttpClient`.

2. I'm well aware they are different products, and I discussed with some of the members of those projects concerns around this exact problem. But this is again a chance for a framework to make life easier, I would rather have my code isolated if possible from declaring everything again if I needed to change from lite -> standard. 

My suggestion is: Let us refactor the code to be able to allow an entry of lite in a way that avoids duplication, and provides a seamless way for users to use. Configurations will be different, they have to. Annotations may have a different property depending on your choice (and even that is debatable with the "endpoint" on the full, could be used as an alias for region).

I debated using endpoint on the raw way, as all apis area always "<region>-<zone>.pubsub.googleapis.com:443", in the end I decided not infer the endpoint as it could change. I'll check with the pubsub team if we are on the safe side by inferring it, and parsing an endpoint into a region/zone is a very simple regex to use and allow portability across environments.

In the end, I'm just a contributor too, but I have been fostering the project and working with Google's internal teams to make sure we are using the APIs on the best way possible.

**Comment by jacobmims** [comment]
Thank you for your response. I think we're in general agreement on eliminating the code duplication and trying to make it as seamless as is reasonably possible. I'll take a crack at refactoring things next week with your concerns in mind. I'm happy to iterate on it a few times to get to a place where we can reach a consensus.

**Comment by viniciusccarvalho** [comment]
@jacobmims please let us refactor the code into separate modules, it will make the PR much simpler to merge and you can start from there.

**Comment by jacobmims** [comment]
@viniciusccarvalho I moved the feature to a separate module and will continue from there, if there are no objections. I had kept it isolated in a separate package, so the refactor was fairly straightforward.

**Comment by viniciusccarvalho** [comment]
@jacobmims As said before, please wait until we make internal changes, work started here: https://github.com/micronaut-projects/micronaut-gcp/tree/pubsublite but I want to hold anything before I get to talk to the pubsub team internally at Google. Anything we do until then is subject to not be aproved.

No need for a new module, we just need a new Factory as the code shows, we may end up with different annotations for the client and listener, but reusing everything else but the factory that either creates: `com.google.cloud.pubsub.v1.Publisher` or `com.google.cloud.pubsublite.cloudpubsub.Publisher`.

I can't work on this until Thursday at best, so please wait for the changes to be merged into master

**Comment by jacobmims** [comment]
Apologies, I misunderstood.

> please let us refactor the code into separate modules, it will make the PR much simpler to merge and you can start from there.

I thought that the first step was to refactor the code to a separate module, then we would start trimming and refactoring. Let me know when I start up again.

**Comment by graemerocher** [comment]
@jacobmims @viniciusccarvalho any update on this?

**Comment by viniciusccarvalho** [comment]
@graemerocher I sat down with the pubsub team last week of December, and there are a few things I had to change, I'm out until the end of this week but will send the final version next week. This PR IMO the way it is should not be merged, I think the best course now is to close the PR and wait for the new refactoring of the base classes.

**Comment by jacobmims** [comment]
> @graemerocher I sat down with the pubsub team last week of December, and there are a few things I had to change, I'm out until the end of this week but will send the final version next week. This PR IMO the way it is should not be merged, I think the best course now is to close the PR and wait for the new refactoring of the base classes.

Ok. I'll close off this implementation, since it sounds like there will be significant changes to the client library.
